AssertionError (lost uuid) in mixed cluster test

This test run failed due to a tripped assertion in one of the 7.0.1 nodes (#86887) but then one of the 7.17.4 nodes also fell over for apparently-unrelated reasons:

```
[2022-05-18T12:15:22,564][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [v7.0.1-0] fatal error in thread [elasticsearch[v7.0.1-0][snapshot][T#3]], exiting
java.lang.AssertionError: lost uuid JCGi3zjyQgytAyClMFs3Gg
        at org.elasticsearch.repositories.RepositoryData.snapshotsToXContent(RepositoryData.java:693) ~[elasticsearch-7.17.4-SNAPSHOT.jar:7.17.4-SNAPSHOT]
        at org.elasticsearch.repositories.RepositoryData.snapshotsToXContent(RepositoryData.java:648) ~[elasticsearch-7.17.4-SNAPSHOT.jar:7.17.4-SNAPSHOT]
        at org.elasticsearch.repositories.blobstore.BlobStoreRepository.lambda$writeIndexGen$65(BlobStoreRepository.java:2323) ~[elasticsearch-7.17.4-SNAPSHOT.jar:7.17.4-SNAPSHOT]
        at org.elasticsearch.common.blobstore.fs.FsBlobContainer.writeToPath(FsBlobContainer.java:274) ~[elasticsearch-7.17.4-SNAPSHOT.jar:7.17.4-SNAPSHOT]
        at org.elasticsearch.common.blobstore.fs.FsBlobContainer.writeBlob(FsBlobContainer.java:253) ~[elasticsearch-7.17.4-SNAPSHOT.jar:7.17.4-SNAPSHOT]
        at org.elasticsearch.repositories.blobstore.BlobStoreRepository.writeAtomic(BlobStoreRepository.java:2636) ~[elasticsearch-7.17.4-SNAPSHOT.jar:7.17.4-SNAPSHOT]
        at org.elasticsearch.repositories.blobstore.BlobStoreRepository.lambda$writeIndexGen$66(BlobStoreRepository.java:2321) ~[elasticsearch-7.17.4-SNAPSHOT.jar:7.17.4-SNAPSHOT]
        at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:136) ~[elasticsearch-7.17.4-SNAPSHOT.jar:7.17.4-SNAPSHOT]
        at org.elasticsearch.common.util.concurrent.ListenableFuture.notifyListenerDirectly(ListenableFuture.java:113) ~[elasticsearch-7.17.4-SNAPSHOT.jar:7.17.4-SNAPSHOT]
        at org.elasticsearch.common.util.concurrent.ListenableFuture.done(ListenableFuture.java:100) ~[elasticsearch-7.17.4-SNAPSHOT.jar:7.17.4-SNAPSHOT]
        at org.elasticsearch.common.util.concurrent.BaseFuture.set(BaseFuture.java:131) ~[elasticsearch-7.17.4-SNAPSHOT.jar:7.17.4-SNAPSHOT]
        at org.elasticsearch.common.util.concurrent.ListenableFuture.onResponse(ListenableFuture.java:139) ~[elasticsearch-7.17.4-SNAPSHOT.jar:7.17.4-SNAPSHOT]
        at org.elasticsearch.action.StepListener.innerOnResponse(StepListener.java:52) ~[elasticsearch-7.17.4-SNAPSHOT.jar:7.17.4-SNAPSHOT]
        at org.elasticsearch.action.NotifyOnceListener.onResponse(NotifyOnceListener.java:29) ~[elasticsearch-7.17.4-SNAPSHOT.jar:7.17.4-SNAPSHOT]
        at org.elasticsearch.repositories.blobstore.BlobStoreRepository.lambda$writeIndexGen$63(BlobStoreRepository.java:2300) ~[elasticsearch-7.17.4-SNAPSHOT.jar:7.17.4-SNAPSHOT]
        at org.elasticsearch.action.ActionRunnable$2.doRun(ActionRunnable.java:62) ~[elasticsearch-7.17.4-SNAPSHOT.jar:7.17.4-SNAPSHOT]
        at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:777) ~[elasticsearch-7.17.4-SNAPSHOT.jar:7.17.4-SNAPSHOT]
        at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26) ~[elasticsearch-7.17.4-SNAPSHOT.jar:7.17.4-SNAPSHOT]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[?:1.8.0_301]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[?:1.8.0_301]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_301]
        Suppressed: java.lang.IllegalStateException: Failed to close the XContentBuilder
                at org.elasticsearch.xcontent.XContentBuilder.close(XContentBuilder.java:1131) ~[elasticsearch-x-content-7.17.4-SNAPSHOT.jar:7.17.4-SNAPSHOT]
                at org.elasticsearch.repositories.blobstore.BlobStoreRepository.lambda$writeIndexGen$65(BlobStoreRepository.java:2322) ~[elasticsearch-7.17.4-SNAPSHOT.jar:7.17.4-SNAPSHOT]
                at org.elasticsearch.common.blobstore.fs.FsBlobContainer.writeToPath(FsBlobContainer.java:274) ~[elasticsearch-7.17.4-SNAPSHOT.jar:7.17.4-SNAPSHOT]
                at org.elasticsearch.common.blobstore.fs.FsBlobContainer.writeBlob(FsBlobContainer.java:253) ~[elasticsearch-7.17.4-SNAPSHOT.jar:7.17.4-SNAPSHOT]
                at org.elasticsearch.repositories.blobstore.BlobStoreRepository.writeAtomic(BlobStoreRepository.java:2636) ~[elasticsearch-7.17.4-SNAPSHOT.jar:7.17.4-SNAPSHOT]
                at org.elasticsearch.repositories.blobstore.BlobStoreRepository.lambda$writeIndexGen$66(BlobStoreRepository.java:2321) ~[elasticsearch-7.17.4-SNAPSHOT.jar:7.17.4-SNAPSHOT]
                at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:136) ~[elasticsearch-7.17.4-SNAPSHOT.jar:7.17.4-SNAPSHOT]
                at org.elasticsearch.common.util.concurrent.ListenableFuture.notifyListenerDirectly(ListenableFuture.java:113) ~[elasticsearch-7.17.4-SNAPSHOT.jar:7.17.4-SNAPSHOT]
                at org.elasticsearch.common.util.concurrent.ListenableFuture.done(ListenableFuture.java:100) ~[elasticsearch-7.17.4-SNAPSHOT.jar:7.17.4-SNAPSHOT]
                at org.elasticsearch.common.util.concurrent.BaseFuture.set(BaseFuture.java:131) ~[elasticsearch-7.17.4-SNAPSHOT.jar:7.17.4-SNAPSHOT]
                at org.elasticsearch.common.util.concurrent.ListenableFuture.onResponse(ListenableFuture.java:139) ~[elasticsearch-7.17.4-SNAPSHOT.jar:7.17.4-SNAPSHOT]
                at org.elasticsearch.action.StepListener.innerOnResponse(StepListener.java:52) ~[elasticsearch-7.17.4-SNAPSHOT.jar:7.17.4-SNAPSHOT]
                at org.elasticsearch.action.NotifyOnceListener.onResponse(NotifyOnceListener.java:29) ~[elasticsearch-7.17.4-SNAPSHOT.jar:7.17.4-SNAPSHOT]
                at org.elasticsearch.repositories.blobstore.BlobStoreRepository.lambda$writeIndexGen$63(BlobStoreRepository.java:2300) ~[elasticsearch-7.17.4-SNAPSHOT.jar:7.17.4-SNAPSHOT]
                at org.elasticsearch.action.ActionRunnable$2.doRun(ActionRunnable.java:62) ~[elasticsearch-7.17.4-SNAPSHOT.jar:7.17.4-SNAPSHOT]
                at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:777) ~[elasticsearch-7.17.4-SNAPSHOT.jar:7.17.4-SNAPSHOT]
                at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26) ~[elasticsearch-7.17.4-SNAPSHOT.jar:7.17.4-SNAPSHOT]
                at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[?:1.8.0_301]
                at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[?:1.8.0_301]
                at java.lang.Thread.run(Thread.java:748) [?:1.8.0_301]
        Caused by: java.io.IOException: Unclosed object or array found
                at org.elasticsearch.xcontent.json.JsonXContentGenerator.close(JsonXContentGenerator.java:456) ~[elasticsearch-x-content-7.17.4-SNAPSHOT.jar:7.17.4-SNAPSHOT]
                at org.elasticsearch.xcontent.XContentBuilder.close(XContentBuilder.java:1129) ~[elasticsearch-x-content-7.17.4-SNAPSHOT.jar:7.17.4-SNAPSHOT]
                ... 19 more
```

**Build scan:**
https://gradle-enterprise.elastic.co/s/hhxbspduxkiio/tests/:qa:mixed-cluster:v7.0.1%23mixedClusterTest/org.elasticsearch.backwards.MixedClusterClientYamlTestSuiteIT/test%20%7Bp0=search.aggregation%2F230_composite%2FTerms%20source%20from%20sorted%7D

**Reproduction line:**
`./gradlew ':qa:mixed-cluster:v7.0.1#mixedClusterTest' -Dtests.class="org.elasticsearch.backwards.MixedClusterClientYamlTestSuiteIT" -Dtests.method="test {p0=search.aggregation/230_composite/Terms source from sorted}" -Dtests.seed=12FB47E08AFF8223 -Dtests.bwc=true -Dtests.locale=el-CY -Dtests.timezone=Atlantic/Stanley -Druntime.java=8`

**Applicable branches:**
7.17

**Reproduces locally?:**
Didn't try

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.backwards.MixedClusterClientYamlTestSuiteIT&tests.test=test%20%7Bp0%3Dsearch.aggregation/230_composite/Terms%20source%20from%20sorted%7D

--- a/docs/changelog/104354.yaml
+++ b/docs/changelog/104354.yaml
@@ -1,0 +1,6 @@
+pr: 104354
+summary: Fix snapshot version in `createSnapshotLegacy`
+area: Snapshot/Restore
+type: bug
+issues:
+ - 86889

--- a/server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java
+++ b/server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java
@@ -273,100 +273,96 @@ public class SnapshotsService extends AbstractLifecycleComponent implements Clus
         final SnapshotId snapshotId = new SnapshotId(snapshotName, UUIDs.randomBase64UUID()); // new UUID for the snapshot
         Repository repository = repositoriesService.repository(request.repository());
         final Map<String, Object> userMeta = repository.adaptUserMetadata(request.userMetadata());
-        clusterService.submitStateUpdateTask(
-            "create_snapshot [" + snapshotName + ']',
-            new ClusterStateUpdateTask(request.masterNodeTimeout()) {
+        repository.executeConsistentStateUpdate(repositoryData -> new ClusterStateUpdateTask(request.masterNodeTimeout()) {
 
-                private List<String> indices;
+            private List<String> indices;
 
-                private SnapshotsInProgress.Entry newEntry;
+            private SnapshotsInProgress.Entry newEntry;
 
-                @Override
-                public ClusterState execute(ClusterState currentState) {
-                    ensureRepositoryExists(repositoryName, currentState);
-                    validate(repositoryName, snapshotName, currentState);
-                    SnapshotDeletionsInProgress deletionsInProgress = currentState.custom(SnapshotDeletionsInProgress.TYPE);
-                    if (deletionsInProgress != null && deletionsInProgress.hasDeletionsInProgress()) {
-                        throw new ConcurrentSnapshotExecutionException(
-                            repositoryName,
-                            snapshotName,
-                            "cannot snapshot while a snapshot deletion is in-progress in [" + deletionsInProgress + "]"
-                        );
-                    }
-                    ensureNoCleanupInProgress(currentState, repositoryName, snapshotName, "create snapshot");
-                    SnapshotsInProgress snapshots = currentState.custom(SnapshotsInProgress.TYPE);
-                    // Fail if there are any concurrently running snapshots. The only exception to this being a snapshot in INIT state from
-                    // a
-                    // previous master that we can simply ignore and remove from the cluster state because we would clean it up from the
-                    // cluster state anyway in #applyClusterState.
-                    if (snapshots != null
-                        && snapshots.asStream()
-                            .anyMatch(
-                                entry -> (entry.state() == State.INIT && initializingSnapshots.contains(entry.snapshot()) == false) == false
-                            )) {
-                        throw new ConcurrentSnapshotExecutionException(repositoryName, snapshotName, " a snapshot is already running");
-                    }
-                    // Store newSnapshot here to be processed in clusterStateProcessed
-                    indices = Arrays.asList(indexNameExpressionResolver.concreteIndexNames(currentState, request));
-
-                    final List<String> dataStreams = indexNameExpressionResolver.dataStreamNames(
-                        currentState,
-                        request.indicesOptions(),
-                        request.indices()
+            @Override
+            public ClusterState execute(ClusterState currentState) {
+                ensureRepositoryExists(repositoryName, currentState);
+                validate(repositoryName, snapshotName, currentState);
+                SnapshotDeletionsInProgress deletionsInProgress = currentState.custom(SnapshotDeletionsInProgress.TYPE);
+                if (deletionsInProgress != null && deletionsInProgress.hasDeletionsInProgress()) {
+                    throw new ConcurrentSnapshotExecutionException(
+                        repositoryName,
+                        snapshotName,
+                        "cannot snapshot while a snapshot deletion is in-progress in [" + deletionsInProgress + "]"
                     );
-
-                    logger.trace("[{}][{}] creating snapshot for indices [{}]", repositoryName, snapshotName, indices);
-                    newEntry = new SnapshotsInProgress.Entry(
-                        new Snapshot(repositoryName, snapshotId),
-                        request.includeGlobalState(),
-                        request.partial(),
-                        State.INIT,
-                        Collections.emptyMap(), // We'll resolve the list of indices when moving to the STARTED state in #beginSnapshot
-                        dataStreams,
-                        threadPool.absoluteTimeInMillis(),
-                        RepositoryData.UNKNOWN_REPO_GEN,
-                        ImmutableOpenMap.of(),
-                        userMeta,
-                        Version.CURRENT
-                    );
-                    initializingSnapshots.add(newEntry.snapshot());
-                    snapshots = SnapshotsInProgress.EMPTY.withAddedEntry(newEntry);
-                    return ClusterState.builder(currentState).putCustom(SnapshotsInProgress.TYPE, snapshots).build();
                 }
-
-                @Override
-                public void onFailure(String source, Exception e) {
-                    logger.warn(() -> new ParameterizedMessage("[{}][{}] failed to create snapshot", repositoryName, snapshotName), e);
-                    if (newEntry != null) {
-                        initializingSnapshots.remove(newEntry.snapshot());
-                    }
-                    newEntry = null;
-                    listener.onFailure(e);
+                ensureNoCleanupInProgress(currentState, repositoryName, snapshotName, "create snapshot");
+                SnapshotsInProgress snapshots = currentState.custom(SnapshotsInProgress.TYPE);
+                // Fail if there are any concurrently running snapshots. The only exception to this being a snapshot in INIT state from a
+                // previous master that we can simply ignore and remove from the cluster state because we would clean it up from the
+                // cluster state anyway in #applyClusterState.
+                if (snapshots != null
+                    && snapshots.asStream()
+                        .anyMatch(
+                            entry -> (entry.state() == State.INIT && initializingSnapshots.contains(entry.snapshot()) == false) == false
+                        )) {
+                    throw new ConcurrentSnapshotExecutionException(repositoryName, snapshotName, " a snapshot is already running");
                 }
+                // Store newSnapshot here to be processed in clusterStateProcessed
+                indices = Arrays.asList(indexNameExpressionResolver.concreteIndexNames(currentState, request));
 
-                @Override
-                public void clusterStateProcessed(String source, ClusterState oldState, final ClusterState newState) {
-                    if (newEntry != null) {
-                        final Snapshot current = newEntry.snapshot();
-                        assert initializingSnapshots.contains(current);
-                        assert indices != null;
-                        beginSnapshot(newState, newEntry, request.partial(), indices, repository, new ActionListener<Snapshot>() {
-                            @Override
-                            public void onResponse(final Snapshot snapshot) {
-                                initializingSnapshots.remove(snapshot);
-                                listener.onResponse(snapshot);
-                            }
+                final List<String> dataStreams = indexNameExpressionResolver.dataStreamNames(
+                    currentState,
+                    request.indicesOptions(),
+                    request.indices()
+                );
 
-                            @Override
-                            public void onFailure(final Exception e) {
-                                initializingSnapshots.remove(current);
-                                listener.onFailure(e);
-                            }
-                        });
-                    }
+                logger.trace("[{}][{}] creating snapshot for indices [{}]", repositoryName, snapshotName, indices);
+                newEntry = new SnapshotsInProgress.Entry(
+                    new Snapshot(repositoryName, snapshotId),
+                    request.includeGlobalState(),
+                    request.partial(),
+                    State.INIT,
+                    Collections.emptyMap(), // We'll resolve the list of indices when moving to the STARTED state in #beginSnapshot
+                    dataStreams,
+                    threadPool.absoluteTimeInMillis(),
+                    RepositoryData.UNKNOWN_REPO_GEN,
+                    ImmutableOpenMap.of(),
+                    userMeta,
+                    minCompatibleVersion(currentState.nodes().getMinNodeVersion(), repositoryData, null)
+                );
+                initializingSnapshots.add(newEntry.snapshot());
+                snapshots = SnapshotsInProgress.EMPTY.withAddedEntry(newEntry);
+                return ClusterState.builder(currentState).putCustom(SnapshotsInProgress.TYPE, snapshots).build();
+            }
+
+            @Override
+            public void onFailure(String source, Exception e) {
+                logger.warn(() -> new ParameterizedMessage("[{}][{}] failed to create snapshot", repositoryName, snapshotName), e);
+                if (newEntry != null) {
+                    initializingSnapshots.remove(newEntry.snapshot());
+                }
+                newEntry = null;
+                listener.onFailure(e);
+            }
+
+            @Override
+            public void clusterStateProcessed(String source, ClusterState oldState, final ClusterState newState) {
+                if (newEntry != null) {
+                    final Snapshot current = newEntry.snapshot();
+                    assert initializingSnapshots.contains(current);
+                    assert indices != null;
+                    beginSnapshot(newState, newEntry, request.partial(), indices, repository, new ActionListener<Snapshot>() {
+                        @Override
+                        public void onResponse(final Snapshot snapshot) {
+                            initializingSnapshots.remove(snapshot);
+                            listener.onResponse(snapshot);
+                        }
+
+                        @Override
+                        public void onFailure(final Exception e) {
+                            initializingSnapshots.remove(current);
+                            listener.onFailure(e);
+                        }
+                    });
                 }
             }
-        );
+        }, "create_snapshot [" + snapshotName + ']', listener::onFailure);
     }
 
     /**

